--- origsrc/jamvm-1.5.4/configure.ac	2014-08-03 16:51:42.451080800 -0500
+++ src/jamvm-1.5.4/configure.ac	2014-08-03 16:43:45.000000000 -0500
@@ -32,11 +32,13 @@ case "$host" in
 i[[3456]]86-*-linux*) host_cpu=i386 host_os=linux ;;
 i[[3456]]86-*-kfreebsd*) host_cpu=i386 host_os=linux ;;
 i[[3456]]86-*-darwin*) host_cpu=i386 host_os=darwin ;;
+i[[3456]]86-*-cygwin*) host_cpu=i386 host_os=cygwin libdl_needed=no ;;
 arm*-*-darwin*) host_cpu=arm host_os=darwin libdl_needed=no ;;
 i386-*-openbsd*) host_os=bsd libdl_needed=no ;;
 i386-*-freebsd*) host_os=bsd libdl_needed=no ;;
 i386-*-solaris*) host_cpu=x86 host_os=solaris ;;
 x86_64-*-linux*) host_os=linux ;;
+x86_64-*-cygwin*) host_os=cygwin libdl_needed=no ;;
 hppa*-*-linux*) host_cpu=parisc host_os=linux ;;
 mipsel-*-linux*) host_cpu=mips host_os=linux ;;
 x86_64-*-openbsd*) host_os=bsd libdl_needed=no ;;
@@ -229,6 +231,7 @@ AC_PROG_CC
 AM_PROG_AS
 
 AC_CHECK_PROGS(JAVAC, ecj jikes "gcj -C" javac)
+PKG_PROG_PKG_CONFIG
 
 dnl Checks for libraries.
 
@@ -248,7 +251,7 @@ if test "$enable_zip" != no; then
 fi
 
 if test "$enable_ffi" != no; then
-    AC_CHECK_LIB(ffi,ffi_call,,AC_MSG_ERROR(libffi is missing))
+    PKG_CHECK_MODULES(FFI,libffi,,AC_MSG_ERROR(libffi is missing))
 fi
 
 dnl Checks for header files.
@@ -259,10 +262,6 @@ if test "$enable_zip" != no; then
     AC_CHECK_HEADER(zlib.h,,AC_MSG_ERROR(zlib.h is missing))
 fi
 
-if test "$enable_ffi" != no; then
-    AC_CHECK_HEADER(ffi.h,,AC_MSG_ERROR(ffi.h is missing))
-fi
-
 if test "$enable_zip" != no; then
     AC_DEFINE([USE_ZIP],1,[use zip])
     use_zip_yes=
@@ -277,6 +276,8 @@ AC_SUBST(use_zip_no)
 
 if test "$enable_ffi" != no; then
     AC_DEFINE([USE_FFI],1,[use FFI])
+    CFLAGS="$CFLAGS $FFI_CFLAGS"
+    LIBS="$LIBS $FFI_LIBS"
 fi
 
 dnl Checks for typedefs, structures, and compiler characteristics.
@@ -322,6 +323,7 @@ AC_CONFIG_FILES(
     src/os/linux/Makefile \
     src/os/darwin/Makefile \
     src/os/bsd/Makefile \
+    src/os/cygwin/Makefile \
     src/os/solaris/Makefile \
     src/os/solaris/x86/Makefile \
     src/os/linux/powerpc/Makefile \
@@ -338,6 +340,8 @@ AC_CONFIG_FILES(
     src/os/bsd/i386/Makefile \
     src/os/bsd/x86_64/Makefile \
     src/os/bsd/sparc/Makefile \
+    src/os/cygwin/i386/Makefile \
+    src/os/cygwin/x86_64/Makefile \
     lib/Makefile \
     lib/java/Makefile \
     lib/java/lang/Makefile \
--- origsrc/jamvm-1.5.4/src/Makefile.am	2014-08-03 16:51:42.451080800 -0500
+++ src/jamvm-1.5.4/src/Makefile.am	2014-08-03 16:53:51.281258600 -0500
@@ -39,10 +39,12 @@ libcore_la_SOURCES = alloc.c alloc.h cas
 jamvm_SOURCES = jam.c
 libjvm_la_SOURCES =
 
-jamvm_LDADD = libcore.la
+jamvm_LDADD = libjvm.la
 libjvm_la_LIBADD = libcore.la
 libcore_la_LIBADD = interp/libinterp.la os/@os@/@arch@/libnative.la os/@os@/libos.la
 
+libjvm_la_LDFLAGS = -avoid-version -no-undefined
+
 AM_CPPFLAGS = -I$(top_srcdir)/src/interp/engine
 
 DISTCLEANFILES = arch.h
--- origsrc/jamvm-1.5.4/src/os/Makefile.am	2014-08-03 16:51:42.461080800 -0500
+++ src/jamvm-1.5.4/src/os/Makefile.am	2014-08-03 16:43:45.000000000 -0500
@@ -20,5 +20,5 @@
 ##
 
 SUBDIRS = @os@
-DIST_SUBDIRS = linux darwin bsd solaris
+DIST_SUBDIRS = linux darwin bsd solaris cygwin
 
--- origsrc/jamvm-1.5.4/src/os/cygwin/Makefile.am	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.4/src/os/cygwin/Makefile.am	2014-08-03 16:43:45.000000000 -0500
@@ -0,0 +1,28 @@
+##
+## Copyright (C) 2003, 2004, 2005, 2006, 2007
+## Robert Lougher <rob@lougher.org.uk>.
+##
+## This file is part of JamVM.
+##
+## This program is free software; you can redistribute it and/or
+## modify it under the terms of the GNU General Public License
+## as published by the Free Software Foundation; either version 2,
+## or (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program; if not, write to the Free Software
+## Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+##
+
+SUBDIRS = @arch@
+DIST_SUBDIRS = i386 x86_64
+
+noinst_LTLIBRARIES = libos.la
+libos_la_SOURCES = os.c
+
+AM_CPPFLAGS = -I$(top_builddir)/src
--- origsrc/jamvm-1.5.4/src/os/cygwin/i386/Makefile.am	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.4/src/os/cygwin/i386/Makefile.am	2014-08-03 16:43:45.000000000 -0500
@@ -0,0 +1,25 @@
+##
+## Copyright (C) 2003, 2004, 2005, 2006, 2007
+## Robert Lougher <rob@lougher.org.uk>.
+##
+## This file is part of JamVM.
+##
+## This program is free software; you can redistribute it and/or
+## modify it under the terms of the GNU General Public License
+## as published by the Free Software Foundation; either version 2,
+## or (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program; if not, write to the Free Software
+## Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+##
+
+noinst_LTLIBRARIES = libnative.la
+libnative_la_SOURCES = init.c
+
+AM_CPPFLAGS = -I$(top_builddir)/src
--- origsrc/jamvm-1.5.4/src/os/cygwin/i386/init.c	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.4/src/os/cygwin/i386/init.c	2014-08-03 18:11:08.631450700 -0500
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 2003, 2004, 2006, 2007
+ * Robert Lougher <rob@lougher.org.uk>.
+ *
+ * This file is part of JamVM.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+void setDoublePrecision(void) {
+}
+
+void initialisePlatform(void) {
+}
--- origsrc/jamvm-1.5.4/src/os/cygwin/os.c	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.4/src/os/cygwin/os.c	2014-08-03 16:51:34.271069500 -0500
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008
+ * Robert Lougher <rob@lougher.org.uk>.
+ *
+ * This file is part of JamVM.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <dlfcn.h>
+#include <unistd.h>
+#include <pthread.h>
+
+#include "../../jam.h"
+
+void *nativeStackBase() {
+    pthread_attr_t attr;
+    void *addr;
+    size_t size;
+
+    pthread_getattr_np(pthread_self(), &attr);
+    pthread_attr_getstack(&attr, &addr, &size);
+
+    return addr+size;
+}
+
+int nativeAvailableProcessors() {
+    return sysconf(_SC_NPROCESSORS_ONLN);
+}
+
+char *nativeLibPath() {
+    return getenv("PATH");
+}
+
+void *nativeLibOpen(char *path) {
+    return dlopen(path, RTLD_LAZY);
+}
+
+void nativeLibClose(void *handle) {
+    dlclose(handle);
+}
+
+void *nativeLibSym(void *handle, char *symbol) {
+    return dlsym(handle, symbol);
+}
+
+char *nativeLibError(void) {
+    return dlerror();
+}
+
+char *nativeLibMapName(char *name) {
+   char *buff = sysMalloc(strlen(name) + sizeof("cyg.dll") + 1);
+
+   sprintf(buff, "cyg%s.dll", name);
+   return buff;
+}
--- origsrc/jamvm-1.5.4/src/os/cygwin/x86_64/Makefile.am	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.4/src/os/cygwin/x86_64/Makefile.am	2014-08-03 16:43:45.000000000 -0500
@@ -0,0 +1,25 @@
+##
+## Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008
+## Robert Lougher <rob@lougher.org.uk>.
+##
+## This file is part of JamVM.
+##
+## This program is free software; you can redistribute it and/or
+## modify it under the terms of the GNU General Public License
+## as published by the Free Software Foundation; either version 2,
+## or (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program; if not, write to the Free Software
+## Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+##
+
+noinst_LTLIBRARIES = libnative.la
+libnative_la_SOURCES = init.c
+
+AM_CPPFLAGS = -I$(top_builddir)/src
--- origsrc/jamvm-1.5.4/src/os/cygwin/x86_64/init.c	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.4/src/os/cygwin/x86_64/init.c	2014-08-03 18:11:15.567831500 -0500
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 2003, 2004, 2006, 2007
+ * Robert Lougher <rob@lougher.org.uk>.
+ *
+ * This file is part of JamVM.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+void setDoublePrecision(void) {
+}
+
+void initialisePlatform(void) {
+}
