--- origsrc/jamvm-1.5.3/configure.ac	2009-04-13 20:05:57.000000000 -0500
+++ src/jamvm-1.5.3/configure.ac	2009-11-09 01:40:16.016497300 -0600
@@ -32,6 +32,7 @@ case "$host" in
 i[[3456]]86-*-linux*) host_cpu=i386 host_os=linux ;;
 i[[3456]]86-*-kfreebsd*) host_cpu=i386 host_os=linux ;;
 i[[3456]]86-*-darwin*) host_cpu=i386 host_os=darwin ;;
+i[[3456]]86-*-cygwin*) host_cpu=i386 host_os=cygwin libdl_needed=no ;;
 arm*-*-darwin*) host_cpu=arm host_os=darwin libdl_needed=no ;;
 i386-*-openbsd*) host_os=bsd libdl_needed=no ;;
 i386-*-freebsd*) host_os=bsd libdl_needed=no ;;
@@ -307,6 +308,7 @@ AC_CONFIG_FILES(
     src/os/linux/Makefile \
     src/os/darwin/Makefile \
     src/os/bsd/Makefile \
+    src/os/cygwin/Makefile \
     src/os/solaris/Makefile \
     src/os/solaris/x86/Makefile \
     src/os/linux/powerpc/Makefile \
@@ -322,6 +324,7 @@ AC_CONFIG_FILES(
     src/os/bsd/arm/Makefile \
     src/os/bsd/i386/Makefile \
     src/os/bsd/x86_64/Makefile \
+    src/os/cygwin/i386/Makefile \
     lib/Makefile \
     lib/java/Makefile \
     lib/java/lang/Makefile \
--- origsrc/jamvm-1.5.3/src/Makefile.am	2009-01-02 02:14:14.000000000 -0600
+++ src/jamvm-1.5.3/src/Makefile.am	2009-11-09 01:46:04.567905100 -0600
@@ -42,6 +42,8 @@ jamvm_LDADD = libcore.la
 libjvm_la_LIBADD = libcore.la
 libcore_la_LIBADD = interp/libinterp.la os/@os@/@arch@/libnative.la os/@os@/libos.la
 
+libjvm_la_LDFLAGS = -no-undefined
+
 AM_CPPFLAGS = -I$(top_srcdir)/src/interp/engine
 
 DISTCLEANFILES = arch.h
--- origsrc/jamvm-1.5.3/src/os/Makefile.am	2008-12-06 16:05:08.000000000 -0600
+++ src/jamvm-1.5.3/src/os/Makefile.am	2009-11-09 01:42:01.940682000 -0600
@@ -20,5 +20,5 @@
 ##
 
 SUBDIRS = @os@
-DIST_SUBDIRS = linux darwin bsd solaris
+DIST_SUBDIRS = linux darwin bsd cygwin solaris
 
--- origsrc/jamvm-1.5.3/src/os/cygwin/Makefile.am	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.3/src/os/cygwin/Makefile.am	2009-11-09 01:42:30.441931700 -0600
@@ -0,0 +1,28 @@
+##
+## Copyright (C) 2003, 2004, 2005, 2006, 2007
+## Robert Lougher <rob@lougher.org.uk>.
+##
+## This file is part of JamVM.
+##
+## This program is free software; you can redistribute it and/or
+## modify it under the terms of the GNU General Public License
+## as published by the Free Software Foundation; either version 2,
+## or (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program; if not, write to the Free Software
+## Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+##
+
+SUBDIRS = @arch@
+DIST_SUBDIRS = i386
+
+noinst_LTLIBRARIES = libos.la
+libos_la_SOURCES = os.c
+
+AM_CPPFLAGS = -I$(top_builddir)/src
--- origsrc/jamvm-1.5.3/src/os/cygwin/i386/Makefile.am	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.3/src/os/cygwin/i386/Makefile.am	2009-11-09 01:41:51.285863400 -0600
@@ -0,0 +1,25 @@
+##
+## Copyright (C) 2003, 2004, 2005, 2006, 2007
+## Robert Lougher <rob@lougher.org.uk>.
+##
+## This file is part of JamVM.
+##
+## This program is free software; you can redistribute it and/or
+## modify it under the terms of the GNU General Public License
+## as published by the Free Software Foundation; either version 2,
+## or (at your option) any later version.
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program; if not, write to the Free Software
+## Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+##
+
+noinst_LTLIBRARIES = libnative.la
+libnative_la_SOURCES = init.c dll_md.c
+
+AM_CPPFLAGS = -I$(top_builddir)/src
--- origsrc/jamvm-1.5.3/src/os/cygwin/i386/dll_md.c	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.3/src/os/cygwin/i386/dll_md.c	2009-11-09 01:41:51.270263400 -0600
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2003, 2004, 2005, 2006, 2007
+ * Robert Lougher <rob@lougher.org.uk>.
+ *
+ * This file is part of JamVM.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "../../../jam.h"
+
+#ifndef USE_FFI
+#include <string.h>
+#include "../../../sig.h"
+
+#define RET_VOID    0
+#define RET_DOUBLE  1
+#define RET_LONG    2
+#define RET_FLOAT   3
+#define RET_DFLT    4
+
+int nativeExtraArg(MethodBlock *mb) {
+    int len = strlen(mb->type);
+    if(mb->type[len-2] == ')')
+        switch(mb->type[len-1]) {
+            case 'V':
+                return RET_VOID;
+            case 'D':
+                return RET_DOUBLE;
+            case 'J':
+                return RET_LONG;
+            case 'F':
+                return RET_FLOAT;
+        }
+
+    return RET_DFLT;
+}
+
+u4 *callJNIMethod(void *env, Class *class, char *sig, int ret_type, u4 *ostack, unsigned char *f, int args) {
+    u4 *opntr = ostack + args;
+    int i;
+
+    for(i = 0; i < args; i++)
+        asm volatile ("pushl %0" :: "m" (*--opntr) : "sp");
+
+    if(class) {
+        asm volatile ("pushl %0" :: "m" (class) : "sp");
+	args++;
+    }
+
+    asm volatile ("pushl %0" :: "m" (env) : "sp");
+
+    switch(ret_type) {
+        case RET_VOID:
+            (*(void (*)())f)();
+            break;
+
+        case RET_DOUBLE:
+            *(double*)ostack = (*(double (*)())f)();
+            ostack += 2;
+            break;
+
+        case RET_LONG:
+            *(long long*)ostack = (*(long long (*)())f)();
+            ostack += 2;
+            break;
+
+        case RET_FLOAT:
+            *(float*)ostack = (*(float (*)())f)();
+            ostack++;
+            break;
+
+        default:
+            *ostack++ = (*(u4 (*)())f)();
+            break;
+    }
+
+    asm volatile ("addl %0,%%esp" :: "r" ((args + 1) * sizeof(u4)) : "cc", "sp");
+    return ostack;
+}
+#endif
--- origsrc/jamvm-1.5.3/src/os/cygwin/i386/init.c	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.3/src/os/cygwin/i386/init.c	2009-11-09 02:31:33.517100600 -0600
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) 2003, 2004, 2006, 2007
+ * Robert Lougher <rob@lougher.org.uk>.
+ *
+ * This file is part of JamVM.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+void setDoublePrecision() {
+}
+
+void initialisePlatform() {
+}
--- origsrc/jamvm-1.5.3/src/os/cygwin/os.c	1969-12-31 18:00:00.000000000 -0600
+++ src/jamvm-1.5.3/src/os/cygwin/os.c	2009-11-09 01:44:48.392972300 -0600
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2003, 2004, 2005, 2006, 2007, 2008
+ * Robert Lougher <rob@lougher.org.uk>.
+ *
+ * This file is part of JamVM.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <dlfcn.h>
+#include <unistd.h>
+
+#define __USE_GNU
+#include <pthread.h>
+
+#include "../../jam.h"
+
+void *nativeStackBase() {
+	return NULL;
+}
+
+int nativeAvailableProcessors() {
+	return sysconf(_SC_NPROCESSORS_ONLN);
+}
+
+char *nativeLibPath() {
+    return getenv("PATH");
+}
+
+void *nativeLibOpen(char *path) {
+    return dlopen(path, RTLD_LAZY);
+}
+
+void nativeLibClose(void *handle) {
+    dlclose(handle);
+}
+
+void *nativeLibSym(void *handle, char *symbol) {
+    return dlsym(handle, symbol);
+}
+
+char *nativeLibMapName(char *name) {
+   char *buff = sysMalloc(strlen(name) + sizeof("cyg.dll") + 1);
+
+   sprintf(buff, "cyg%s.dll", name);
+   return buff;
+}
