--- origsrc/jamvm-1.5.3/src/jam.c	2009-01-31 02:13:37.000000000 -0600
+++ src/jamvm-1.5.3/src/jam.c	2012-01-09 04:30:45.900787200 -0600
@@ -140,7 +140,7 @@ int parseCommandLine(int argc, char *arg
     int status = 1;
     int i;
 
-    Property props[argc-1];
+    Property props[argc];
     int props_count = 0;
 
     for(i = 1; i < argc; i++) {
@@ -249,6 +249,12 @@ int parseCommandLine(int argc, char *arg
             props[props_count].key = key;
             props[props_count++].value = pntr;
 
+            if(strcmp(key, "java.awt.headless") == 0 &&
+               strcmp(pntr, "true") == 0) {
+                props[props_count].key = strcpy(sysMalloc(12), "awt.toolkit");
+                props[props_count++].value = strcpy(sysMalloc(43), "gnu.java.awt.peer.headless.HeadlessToolkit");
+            }
+
         } else if(strcmp(argv[i], "-classpath") == 0 ||
                   strcmp(argv[i], "-cp") == 0) {
 
